src/viam/sdk/common/client_helper.hpp

```diff
@@ -63,21 +77,52 @@ class ClientHelper {
         client_helper_details::errorHandlerReturnedUnexpectedly(result);
     }
 
+    // A version of invoke for gRPC calls returning `(stream ResponseType)`.
+    // ResponseHandlerCallable will be called for every response in the reader, and should return
+    // false to indicate it is no longer interested in the stream.
+    template <typename ResponseHandlerCallable>
+    auto invoke_stream(ResponseHandlerCallable rhc) {
+        *request_.mutable_name() = client_->name();
+        ClientContext ctx;
+
+        auto reader = (stub_->*pfn_)(ctx, request_);
```

Comment:
Looking more carefully at [grpc::ClientReaderInterface](https://grpc.github.io/grpc/cpp/classgrpc_1_1_client_reader_interface.html), I see that there is a `Finish` call with seems related to error handling / status of the RPC. I wonder if we were failing to interact with that properly in the existing code, and whether this helper is the opportunity to do the right thing?

Reply:
I think you're right, the [docs](https://grpc.github.io/grpc/cpp/classgrpc_1_1internal_1_1_client_streaming_interface.html#a567f213cc0a5df2a03becda3e5711e25) suggest to me that this should be called after the `while` loop. If we drained the reader it will return a status immediately, and if our stream handler terminated the reading it will flush the remaining messages and then return a status.

Reply:
This was surprisingly not forthcoming info, but it looks like `TryCancel` is the way to do it, and then a call to `Finish` should return an OK or Cancelled status, see eg 

https://groups.google.com/g/grpc-io/c/R0NTqKaHLdE?pli=1
https://github.com/grpc/grpc/issues/17292#issuecomment-460773219
https://stackoverflow.com/a/54560700

Reply:
I'm learning this here with you, so I have a few more questions:
- Is it possible for (stub->*pfn) to return an empty `unique_ptr` such that `reader` is not valid? If that happens, is there an error to be obtained from the `ClientContext` in some way?
- I wonder if (and sorry, this will mean undoing an earlier suggestion of mine) the case of `reader->Read(&response)` returning false and `rhc` returning `false` now need to be handled differently. In particular, if `reader->Read` returns false, I believe it means no more messages are coming, so a `Cancel` may not be needed (but an error check is!), whereas if `rhc` returns `false` it means the users callback no longer wants more messages, so a `TryCancel` is probably needed.

Reply:
To the first point, this was again not forthcoming, but it looks like no: the stub calls its own `StreamEventsRaw` method, which then calls `internal::ClientReaderFactory<R>::Create`, and [that call](https://github.com/grpc/grpc/blob/master/include/grpcpp/support/sync_stream.h#L170) is always a `new` construction

Second point, I mulled that over for a bit and found someone asking in more detail but not really getting a helpful reply, because imo this ticket points out an ambiguity/absence in docs:
https://github.com/grpc/grpc/issues/22839
Anyway I'll spare you all the permutations I was considering but yes maybe it's better to revert to separate condition checking. I *suspect* that it's fine to call `TryCancel` when the stream already finished, but we may as well only do it if the reader is dry anyway.

Reply:
OK, I think the logic looks pretty solid to me now. Resolving.